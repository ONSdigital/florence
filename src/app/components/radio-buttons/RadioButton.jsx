import React, { Component } from "react";
import PropTypes from "prop-types";

const propTypes = {
    id: PropTypes.string.isRequired,
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    group: PropTypes.string.isRequired,
    onChange: PropTypes.func,
    checked: PropTypes.bool,
    inline: PropTypes.bool,
    disabled: PropTypes.bool,
    subLabel: PropTypes.string,
};

export default class RadioButton extends Component {
    constructor(props) {
        super(props);

        this.state = {
            focused: false,
            checked: props.checked,
        };
    }

    handleChange = event => {
        const checked = event.target.checked;
        const id = this.props.id;
        const value = this.props.value;
        const onChange = this.props.onChange;
        if (onChange) {
            onChange({ id, checked, value });
        }
    };

    handleFocus = () => {
        this.setState({ focused: true });
    };

    handleBlur = () => {
        this.setState({ focused: false });
    };

    render() {
        return (
            <div className={"radio" + (this.props.inline ? " radio--inline" : "")}>
                <input
                    className={"radio__input" + (this.props.checked ? " selected" : "") + (this.state.focused ? " focused" : "")}
                    type="radio"
                    name={this.props.group}
                    checked={this.props.checked}
                    value={this.props.value}
                    id={this.props.id}
                    onFocus={this.handleFocus}
                    onBlur={this.handleBlur}
                    onChange={this.handleChange}
                    disabled={this.props.disabled}
                />

                <label
                    className={
                        "radio__label" +
                        (this.props.checked ? " radio__label--selected" : "") +
                        (this.state.focused ? " radio__label--focused" : "") +
                        (this.props.disabled ? " radio__label--disabled" : "")
                    }
                    htmlFor={this.props.id}
                >
                    {this.props.label}{" "}
                    {this.props.subLabel ? (
                        <span className="sub-label">
                            <br />
                            {this.props.subLabel}
                        </span>
                    ) : null}
                </label>
            </div>
        );
    }
}

RadioButton.propTypes = propTypes;
