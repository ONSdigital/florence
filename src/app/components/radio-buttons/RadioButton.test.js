import React from 'react';
import RadioButton from './RadioButton';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

const defaultProps = {
    id: "test-radio",
    value: "test-value",
    label: "Test radio",
    group: "test-group",
    onChange: ()=>{},
    checked: false,
    inline: false
};

test("Radio button matches stored snapshot", () => {

    const component = renderer.create(
        <RadioButton {...defaultProps} />
    );
    expect(component.toJSON()).toMatchSnapshot();
});

test("Handle change passes correct value to parent onChange method", () => {
    let passedToParent = {};
    const mockedOnChange = (event) => {passedToParent = event;};
    const component = shallow(
        <RadioButton {...defaultProps} checked={false} onChange={mockedOnChange}/>
    );
    expect(component.update().state().checked).toBe(false);
    component.instance().handleChange({target: {checked: true}});
    expect(passedToParent).toMatchObject({
        id: "test-radio",
        checked: true,
        value: "test-value"
    });
});

test("Handle focus updates state correctly", () => {
    const component = shallow(
        <RadioButton {...defaultProps} />
    );
    expect(component.update().state().focused).toBe(false);
    component.instance().handleFocus();
    expect(component.update().state().focused).toBe(true);
});

test("Handle blur updates state correctly", () => {
    const component = shallow(
        <RadioButton {...defaultProps} />
    );
    component.setState({ focused: true });
    expect(component.update().state().focused).toBe(true);
    component.instance().handleBlur();
    expect(component.update().state().focused).toBe(false);
});
