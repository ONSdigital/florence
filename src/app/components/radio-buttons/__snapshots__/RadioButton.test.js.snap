// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio button matches stored snapshot 1`] = `
<div
  className="radio"
>
  <input
    checked={false}
    className="radio__input"
    id="test-radio"
    name="test-group"
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    type="radio"
    value="test-value"
  />
  <label
    className="radio__label"
    htmlFor="test-radio"
  >
    Test radio
     
     
  </label>
</div>
`;
